bpf_queue.c

```diff
@@ -526,6 +527,109 @@ bpf_ringbuf_cb(void *vqq, void *vdata, size_t len)
 	return (0);
 }
 
+/*
+ * Lookup where cgroup2 is mounted, if it's not, try to mount it ourselves, if
+ * so, fill unmount_path with the temporary cgroup2 path that must be unmounted
+ * by calling cgroup2_unmount_tmp().
+ */
+static int
+cgroup2_open_fd(char **unmount_path)
+{
+	char	*save_line, *file_buf;
+	char	*line, *start, *end, *path;
+	int	 fd;
+
+	fd = -1;
+	path = NULL;
+
+	if ((file_buf = load_file_path_nostat("/proc/mounts", NULL)) == NULL) {
+		qwarn("load_file_path_nostat /proc/mounts");
+		goto fail;
+	}
+
+	for (line = strtok_r(file_buf, " ", &save_line);
+	     line != NULL;
+	     line = strtok_r(NULL, "\n", &save_line)) {
+		if (strncasecmp(line, "cgroup2", strlen("cgroup2")))
+			continue;
+		if ((start = strchr(line, ' ')) == NULL) {
+			qwarn("no space 1");
+			continue;
+		}
+		start++;
+		if ((end = strchr(start, ' ')) == NULL) {
+			qwarn("no space 2");
+			continue;
+		}
+		path = strndup(start, end - start);
+		break;
+	}
+
+	/*
+	 * No cgroup2 mount found, try mountint it ourselves at
+	 * /tmp/quark_cgroup2_mount.XXXXX.
+	 * If we succeed, we must pass the mounting point up to the caller, so
+	 * it can umount after the probes are loaded.
+	 */
+	if (path == NULL) {
+		char template[] = "/tmp/quark_cgroup2_mount.XXXXXX";
+
+		if ((path = mkdtemp(template)) == NULL) {
+			qwarn("mktemp %s", template);
+			goto fail;
+		}
+		qwarnx("no cgroup2 mount found, will try mounting it "
+		    "ourselves at %s", path);
+
+		path = strdup(path);
+		if (path == NULL) {
+			qwarn("strdup");
+			goto fail;
+		}
+		if (mount(NULL, path, "cgroup2", 0, NULL) == -1) {
+			qwarn("mount %s", path);
+			goto fail;
+		}
+
+		*unmount_path = strdup(path);
```

Comment:
we check for `strdup` failures on line 585. should make them consistent. if it fails (and all other code miraculously works) we will fail to umount the custom cgroup2 mount.

Reply:
I thought about it, it would basically just leave it dangling, while in practice that strdup will never fail, so I just ignored. I can add some print at least